### DISCUSSION TITLE
More information about timeouts in Tasks and PipelineRuns

### DISCUSSION THREAD

**Discussion by pvlltvk** [tektoncd/pipeline#4200] [Q&A]
Hi, everyone!

Can someone please clarify the relationships between the `timeout` field in the `Tasks` spec of a `Pipeline`  and the `timeouts` field in a `PipelineRun`?
I created a `Pipeline` and set the failure timeout for one of the heavy tasks to 24h0m0s. I run a `PipelineRun` for that `Pipeline` (I didn't set the `timeouts` field for this `PipelineRun`) and I got `PipelineRunTimeout` after 1 hour but my heavy task continued to work. 
Is it a correct behaviour?

**Comment by jerop** [comment]
that doesn't sound like the correct behavior to me 

in [TEP-0046: Finally tasks execution post pipelinerun timeout](https://github.com/tektoncd/community/blob/main/teps/0046-finallytask-execution-post-timeout.md#pipeline-timeout), we provided a way for `finally tasks` to be executed post `pipelinerun` timeout -- is your heavy task a `finally task`? 
- if yes, i'd expect it to continue executing 
- if no, i'd expect it to timeout and execute `finally tasks` if any 

but it doesn't seem that this scenario was explicitly discussed (unless i missed something)

cc @tektoncd/core-maintainers @souleb

**Comment by souleb** [comment]
Hi,

Can you provide the yaml to cross check your declaration? I have read the code once again and here a first attempt to explain.

When `timeouts` fields are used (alpha features enabled), 'timeouts.pipeline` sets the maximum allowed duration for execution of the entire pipeline. The sum of individual `timeouts` for `tasks` and `finally` must not exceed this value.

In your case you do not set any timeout for your `pipelinerun`, so the default value of 60m minutes is used.
https://github.com/tektoncd/pipeline/blob/f041ef5d644c72e88689e079a8e3ea96718891fc/pkg/apis/pipeline/v1beta1/pipelinerun_defaults.go#L36

Now why does your `taskrun` still run after the hour? When a `task` is resolved at the `taskrun` creation time, if the resolved `task` set a timeout, then it takes precedence.
https://github.com/tektoncd/pipeline/blob/f041ef5d644c72e88689e079a8e3ea96718891fc/pkg/reconciler/pipelinerun/pipelinerun.go#L1126

The `taskrun` reconciler does not watch `pipelinerun` for timeout event...

This is not the intended behaviour in my opinion. We should maybe check the provided `task` timeout when resolved, but that is after validation so seems too late...or an ownerRef watcher?
